File: 2021/Day01/solution.py
import os 
os.chdir("D:\\Programming\\AdventOfCode\\2021\\Day01\\")

input = [int(x) for x in open('input.txt')]    

# Part 1
sum(x > y for x,y in zip(input[1:], input))


# Part 2 
rolling_sums = [sum(x) for x in list(zip(input, input[1:], input[2:]))]

sum(x > y for x,y in zip(rolling_sums[1:], rolling_sums))

# Part 2 optimised
sum(x > y for x,y in zip(input[3:], input))

File: 2021/Day01/solution.R
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

library(dplyr)      # for lag
library(data.table) # for rolling sum

input <- readLines("input.txt")
input <- as.numeric(input)


# Part 1

increases <- input - lag(input)

sum(increases > 0, na.rm = TRUE)


# Part 2

sliding_sums <- data.table::frollsum(input, n = 3)

increases <- sliding_sums - lag(sliding_sums)

sum(increases > 0, na.rm = TRUE)


# Part 2 optimised
sum(input > lag(input, 3), na.rm = TRUE)
